pyrolite/mineral/normative.py

```diff
@@ -137,95 +133,233 @@ def endmember_decompose(
     modal = renormalise(modal)
     return modal
 
-
-def CIPW_norm(df, form="weight"):
+def CIPW_norm(data, form="weight"):
     """
     Standardised calcuation of estimated mineralogy from bulk rock chemistry.
-    Takes a dataframe of chemistry and creates a dataframe of estimated mineralogy.
-
+    Takes a dataframe of chemistry & creates a dataframe of estimated mineralogy.
+    
+    This is the CIPW norm of Verma et al. (2003)
+    
+    This version only uses major elements
     Parameters
     -----------
-    df : :class:`pandas.DataFrame`
+    df : :class:`p&as.DataFrame`
         Dataframe containing compositions to transform.
-
     Returns
     --------
     :class:`pandas.DataFrame`
-
     Notes
     -----
     This function is currently a stub.
     """
-
-    # devolatilise and renormalise
-
-    # deal with Fe oxidation
-
-    # convert to molar data
-
-    # aggregate like-elements to their most commom member
-    # Ferric : FeO, MnO, NiO, CoO
-    # Calcic : CaO, BaO, SrO
-    # Potassic : K2O, Rb2O, Cs2O
-    # Sodic : Na2O, Li2O
-    # Chromic : Cr2O3, V2O3
-
-    # recalculate 'molecular weigths' for each of these components, and the minerals
-
-    # Zircon
-
+    columns = ['SiO2','TiO2','Al2O3','Fe2O3','FeO','MnO','MgO','CaO','Na2O','K2O','P2O5','CO2','SO3']
+    mol_weights = pd.Series([60.08,79.90,101.96,159.6882,71.844,70.94,40.30,56.08,61.98,94.20,141.94, 44.01, 80.06], index = columns)
+
+    # If want to convert to FeOtotal
+#     data['FeO'] = data['FeO'] + (data['Fe2O3']*(2*71.8444/159.6882))
+#     data['Fe2O3'] = 0
+
+    # Normalise to 100 on anhydrous basis
+    totals = data.sum(axis = 1)
+    data = data.divide(totals/100, axis = 0)
+
+    # Convert to moles
+    data = data.div(mol_weights)
+
+    # Calculate corrected molecular weights for later
+    # When updating for trace elements will need to add more here
+    xFeO = data['FeO']/(data['FeO']+data['MnO'])
+    xMnO = data['MnO']/(data['FeO']+data['MnO'])
+    MWFeOcorr = (xMnO*mol_weights['MnO'])+(xFeO*mol_weights['FeO'])
```

Comment:
For this section, I think it could be cleaned up a little and some precision added thanks to `periodictable`:
```python
############################################################################
# Calculate effective molecular weights for silicate & oxide endmembers
############################################################################
# When updating for trace elements will need to add more here
mineral_mw = {}

# Components (used only for calc) ##########################################
# effective MW of (Fe2+, Mn2+)O
molw_FeO = (xMnO * mol_w["MnO"]) + (xFeO * mol_w["FeO"])
# effective MW of Mn-fayalite
molw_fayalite = (2 * molw_FeO) + pt.formula("SiO2").mass
# effective v of ferrosilite (Fe2+, Mn2+)SiO3
molw_fs = molw_FeO + pt.formula("SiO2").mass
# Endmembers (used in the norm) ############################################
# effective MW of ilmenite (Fe2+, Mn2+)TiO3
molw_il = molw_FeO + pt.formula("TiO2").mass
# effective MW of hypersthene (Mg2+, Fe2+, Mn2+)SiO3
molw_hyp = pt.formula("MgSiO3").mass * MgNo + molw_fs * FeNo
# effective MW of olivine (Mg2+, Fe2+, Mn2+)SiO4
molw_ol = pt.formula("Mg2SiO4").mass * MgNo + molw_fayalite * FeNo
# effective MW of magnetite
molw_mt = molw_FeO_corr + pt.formula("Fe2O3").mass

for n, mw in zip(
    ["ilmenite", "hypersthene", "olivine", "magnetite"],
    [molw_il, molw_hyp, molw_ol, molw_mt],
):
    mineral_mw[n] = mw

# anhydrous apatite (Ca10P6O25)/3 relative to P2O5
mineral_mw["apatite"] = (
    pt.formula("Ca10(PO4)6(OH)2").mass - pt.formula("H2O").mass
) / 3
mineral_mw["halite"] = pt.formula("NaCl").mass  # leave out Cl?
mineral_mw["fluorite"] = pt.formula("CaF2").mass  # leave out F2?
mineral_mw["pyrite"] = pt.formula("FeS2").mass
############################################################################
# Aggregate Mn, Fe
############################################################################
res["FeO"] = res["FeO"] + res["MnO"]
res = res.drop(["MnO"], axis=1)
```
Might want to confirm that the apatite is about correct, I didn't check it relative to your calculation below, but I think it's roughly what's happening?

Also - I was wondering whether you need to include or exclude the halogens here, as they don't form part of the basis we use at the start (similar to using anhydrous apatite)?